### PR TITLE
PR - Kittenservice and Kittens list to dev

### PR DESCRIPTION
**Pull request by carlsavila**
kittenservice implementation with the kitttens list display.

### CODE REVIEW

File: src/app/components/kitten/kitten/kitten.component.scss [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        font-size: 1rem;
+    }
+
+    .pem {
+        color: var(--kittenCard-pem);
+        font-weight: bold;
+    }
+}
+
+/* Medium devices (landscape tablets, 768px and up) */
```
Comment (Benja-san) [inline src/app/components/kitten/kitten/kitten.component.scss]: remove dead code

File: src/app/components/kitten/kitten/kitten.component.scss [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+/* Large devices (laptops/desktops, 992px and up) */
+@media screen and (min-width: 601px) and (max-width: 992px) {
+    /* Kittend Card */
```
Comment (Benja-san) [inline src/app/components/kitten/kitten/kitten.component.scss]: remove dead code

File: src/app/components/kitten/kitten/kitten.component.scss [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+        .pem {
+            color: var(--kittenCard-pem);
+            font-weight: bold;
+        }
+    }
+}
+
+/* Large devices (laptops/desktops, 992px and up) */
+/* Extra large devices (large laptops and desktops, 1200px and up) */
```
Comment (Benja-san) [inline src/app/components/kitten/kitten/kitten.component.scss]: remove dead code

File: src/app/components/kitten/kitten/kitten.component.scss [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+@media screen and (min-width: 992px) {
+    /* Kittend Card */
```
Comment (Benja-san) [inline src/app/components/kitten/kitten/kitten.component.scss]: remove dead code

File: src/app/model/animal/animal.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    constructor(
+        private _name: string,
+        private readonly _breed: string,
+        private readonly _birthday: Date,
+        private _image: string,
+    ) {
+
+        Animal.addKittNumber();
+        this._id = Animal.getKittNumber();
+        console.log("compteur : " + Animal.getKittNumber() +  ". id "+  this.name + " " + this.id);
```
Comment (Benja-san) [inline src/app/model/animal/animal.ts]: remove console log

File: src/app/pages/kittens/kittens.component.html [hunk 6]
```diff
@@ -0,0 +1,40 @@
+<main class="main-grid-container">
+
+    <!-- AVAILABLE kittens section -->
+    <section class="grid-itemList">
+        <h3>Chatons disponibles</h3>
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.html]: page starts with azn H1

File: src/app/pages/kittens/kittens.component.html [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+        <!--<button (click)="pushKittenForTest()"> ajouter</button>-->
+        <div class="grid-itemList-CardsContainer">
+
+            @for (kitten of getServiceKittenAvailableKittens(); track $index) {
+            <div class="grid-itemList-CardContainer">
+                <div>
+                    <a (click)="addAdoptedKittenToUserKittenLis($index)">
+                        <!-- Card @Input -->
+                        <app-kitten [kittenCard]="kitten"></app-kitten>
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.html:9]: and organize H5 tag inside this component to be a logical suit to the previous h in this page

File: src/app/pages/kittens/kittens.component.html [hunk 8]
```diff
@@ -0,0 +1,40 @@
+<main class="main-grid-container">
+
+    <!-- AVAILABLE kittens section -->
+    <section class="grid-itemList">
+        <h3>Chatons disponibles</h3>
+
+        <!--<button (click)="pushKittenForTest()"> ajouter</button>-->
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.html]: remove dead code

File: src/app/pages/kittens/kittens.component.html [hunk 9]
```diff
@@ -0,0 +1,40 @@
+<main class="main-grid-container">
+
+    <!-- AVAILABLE kittens section -->
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.html]: remove dead code

File: src/app/pages/kittens/kittens.component.html [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        <h3>Chatons disponibles</h3>
+
+        <!--<button (click)="pushKittenForTest()"> ajouter</button>-->
+        <div class="grid-itemList-CardsContainer">
+
+            @for (kitten of getServiceKittenAvailableKittens(); track $index) {
+            <div class="grid-itemList-CardContainer">
+                <div>
+                    <a (click)="addAdoptedKittenToUserKittenLis($index)">
+                        <!-- Card @Input -->
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.html]: remove dead code

File: src/app/pages/kittens/kittens.component.html [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+                            <em> et je suis à toi</em> 🥰💞💘
+                        </p>
+                    </a>
+                </div>
+            </div>
+            }
+        </div>
+    </section>
+
+    <!-- USER section - user's kittens list
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.html]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 12]
```diff
@@ -0,0 +1,371 @@
+/****** LIST in main ****/
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+.main-grid-container {
+    text-align: center;
+    background-color: var(--body-HeaderMainFooter-bg);
+    padding: 0;
+    margin: 0;
+    
+    border: var(--main-grid-container-border); // For DEBUGGING pourpose. to remove
+    //border: .1px solid greenyellow; // For DEBUGGING pourpose. to remove
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+
+    /* List Available Kitten */
+    .grid-itemList {
+        box-shadow: var(--global-bsdw-color);
+        border-radius: 10px;
+
+        //border: var(--grid-itemList); // For DEBUGGING pourpose. to remove
+        //border: 5px solid fuchsia; // For DEBUGGING pourpose. to remove
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+
+        .grid-itemList-CardsContainer {
+            border-radius: 10px;
+            background-color: var(--list-CardsContainer-bg-color);
+
+            //border: var(--grid-itemList-CardsContainer-boder); // For DEBUGGING pourpose. to remove
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+        .grid-itemList-CardContainer:hover {
+            width: 22rem;
+            height: fit-content;
+            box-shadow: var(--global-bsdw-color);
+            transition: all 1s linear;
+            border: var(--grid-itemList-CardContainer-hover);
+        }
+    }
+
+    /* User Kitten List */
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+                    width: 22rem;
+                    height: fit-content;
+                    box-shadow: var(--global-bsdw-color);
+                    transition: all 1s linear;
+                    border: var(--grid-itemUser-kittenCard-Container-hover);
+                }
+            }
+        }
+    }
+    /* end list */
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+
+        border: var(--grid-itemForm);
+
+        // For DEBUGGING pourpose. to remove
+        border: 5px solid greenyellow;
+    }
+}
+
+
+ /*** https://www.w3schools.com/css/css_rwd_mediaqueries.asp ***/
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+                    display: flex;
+                    flex-direction: column;
+                    align-items: center;
+                    justify-content: center;
+                    padding: 0;
+                    gap: 2px;
+                }
+            }
+
+            /*end list*/
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+                }
+            }
+        }
+
+
+
+    }
+
+
+/* Medium devices (landscape tablets, 768px and up) */
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+/* Large devices (laptops/desktops, 992px and up) */
+@media screen and (min-width: 601px) and (max-width: 992px) {
+
+    /** liste **/
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 22]
```diff
@@ -0,0 +1,371 @@
...
+                    .grid-itemUser-kittenCard-Container {
+                        padding: .5px;
+                        width: 100%;
+                        max-width: 22rem;
+                    }
+                }
+            }
+        }
+
+        /*end list*/
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+                        width: 20rem;
+                    }
+                }
+            }
+        }
+    }
+
+}
+
+/* Large devices (laptops/desktops, 992px and up) */
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 24]
```diff
@@ -0,0 +1,371 @@
...
+/* Extra large devices (large laptops and desktops, 1200px and up) */
+@media screen and (min-width: 992px) {
+
+    /** List **/
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 25]
```diff
@@ -0,0 +1,371 @@
...
+    // main grid LIST available KITTENS/USER4 KITTENS/FORM NEW KITTEN //
+    .main-grid-container {
+        display: grid;
+        grid-template-areas:
+            'list list user'
+            'list list user'
+            'form form user';
+
+        // For DEBUGGING pourpose. to remove
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 26]
```diff
@@ -0,0 +1,371 @@
...
+        border: 5px solid green;
+
+        // section grid LIST available KITTENS //
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+        .grid-itemList {
+            //padding: .5rem;
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 28]
```diff
@@ -0,0 +1,371 @@
...
+            display: flex;
+            flex-direction: column;
+            align-items: center;
+            justify-content: center;
+            width: 600px;
+            height: 100%;
+
+            // For DEBUGGING pourpose. to remove
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 29]
```diff
@@ -0,0 +1,371 @@
...
+                align-items: center;
+                justify-items: center;
+                position: relative;
+                width: 100%;
+
+                .grid-itemList-CardContainer {
+                    width: 100%;
+                    max-width: 22rem;
+
+                    // For DEBUGGING pourpose. to remove
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 30]
```diff
@@ -0,0 +1,371 @@
...
+                    border: 5px solid dodgerblue;
+                }
+
+            }
+        }
+
+        // section grid USER'S KITTENS //
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 31]
```diff
@@ -0,0 +1,371 @@
...
+            padding: .1rem;
+            display: flex;
+            flex-direction: column;
+            justify-content: flex-start;
+            align-items: center;
+
+            width: 200px;
+            height: 100%;
+
+            // For DEBUGGING pourpose. to remove
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.scss [hunk 32]
```diff
@@ -0,0 +1,371 @@
...
+                        padding: .5px;
+                        width: 100%;
+                        max-width: 22rem;
+                    }
+                }
+            }
+        }
+        /* end list */
+
+        // section grid FORME ADD A KITTEN //
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.scss]: remove dead code

File: src/app/pages/kittens/kittens.component.ts [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+@Component({
+  selector: 'app-kittens',
+  standalone: true,
+  imports: [KittenComponent],
+  templateUrl: './kittens.component.html',
+  styleUrl: './kittens.component.scss'
+})
+export class KittensComponent {
+
+  // KittensService injection (to get, add available kittens and adopted user's kitten(s)).
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: remove dead code

File: src/app/pages/kittens/kittens.component.ts [hunk 34]
```diff
@@ -0,0 +1,63 @@
...
+  private kittenService: KittensService = inject(KittensService);
+
+  getServiceKittenJsonKittens(): void {
+    console.log("Récupération Json Kittens : ")
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: remove console log

File: src/app/pages/kittens/kittens.component.ts [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+  imports: [KittenComponent],
+  templateUrl: './kittens.component.html',
+  styleUrl: './kittens.component.scss'
+})
+export class KittensComponent {
+
+  // KittensService injection (to get, add available kittens and adopted user's kitten(s)).
+  private kittenService: KittensService = inject(KittensService);
+
+  getServiceKittenJsonKittens(): void {
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts:17]: getKittens*

File: src/app/pages/kittens/kittens.component.ts [hunk 36]
```diff
@@ -0,0 +1,63 @@
...
+    console.log("Récupération Json Kittens : ")
+    this.kittenService.getJsonKittensAndPushToAvailableKittens();
+  }
+
+  getServiceKittenAvailableKittens(): Animal[] {
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts:21]: getAvailableKittens*

File: src/app/pages/kittens/kittens.component.ts [hunk 37]
```diff
@@ -0,0 +1,63 @@
...
+    return this.kittenService.getAvailableKittens();
+  }
+
+  // Function to add new kitten received from create-kitten's @Output, using the form
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: rm dead code

File: src/app/pages/kittens/kittens.component.ts [hunk 38]
```diff
@@ -0,0 +1,63 @@
...
+  addNewAvailableKitten(event: Animal) {
+    this.kittenService.addAvailableKitten(event);
+  }
+
+  // Funnction to send the adopted kitten to the user
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: rm dead code

File: src/app/pages/kittens/kittens.component.ts [hunk 39]
```diff
@@ -0,0 +1,63 @@
...
+  addAdoptedKittenToUserKittenLis(kittenIndex: number) {
+
+    console.log("got index ", kittenIndex)
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: remove console log

File: src/app/pages/kittens/kittens.component.ts [hunk 40]
```diff
@@ -0,0 +1,63 @@
...
+    // Searching the adopted kitten by id in available kittens list.
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: rm dead code

File: src/app/pages/kittens/kittens.component.ts [hunk 41]
```diff
@@ -0,0 +1,63 @@
...
+    this.kittenService.getAvailableKittens().find(
+      (kittenItem, index) => {
+
+        if (index === kittenIndex) {
+          console.log("item ", kittenItem, " of index ", index);
+          // Adding found kitten to the user's adopted kitten.
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: rm dead code

File: src/app/pages/kittens/kittens.component.ts [hunk 42]
```diff
@@ -0,0 +1,63 @@
...
+          console.log("kitten found ", kittenItem)
+          this.kittenService.addAdoptedKitten(kittenItem);
+
+          const indexOfItem = this.kittenService.getAvailableKittens().indexOf(kittenItem);
+          // Delete from list by element corresponding to the indexOfItem
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: rm dead code

File: src/app/pages/kittens/kittens.component.ts [hunk 43]
```diff
@@ -0,0 +1,63 @@
...
+
+    console.log("got index ", kittenIndex)
+    // Searching the adopted kitten by id in available kittens list.
+    this.kittenService.getAvailableKittens().find(
+      (kittenItem, index) => {
+
+        if (index === kittenIndex) {
+          console.log("item ", kittenItem, " of index ", index);
+          // Adding found kitten to the user's adopted kitten.
+          console.log("kitten found ", kittenItem)
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts]: remove console log

File: src/app/pages/kittens/kittens.component.ts [hunk 44]
```diff
@@ -0,0 +1,63 @@
...
+          const indexOfItem = this.kittenService.getAvailableKittens().indexOf(kittenItem);
+          // Delete from list by element corresponding to the indexOfItem 
+          this.kittenService.removeAvailableKitten(indexOfItem);
+
+          alert("Et voilà! Pour la vie ! 😀")
+        }
+      })
+  }
+
+  getServiceKittenAdoptedKitten(): Animal[] {
```
Comment (Benja-san) [inline src/app/pages/kittens/kittens.component.ts:45]: getAdoptedKittens

File: src/app/services/kittens.service.ts [hunk 45]
```diff
@@ -0,0 +1,63 @@
+import { Injectable, inject } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { Animal } from '../model/animal/animal';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class KittensService {
+
+  // List available Kittens.
```
Comment (Benja-san) [inline src/app/services/kittens.service.ts]: remove dead code

File: src/app/services/kittens.service.ts [hunk 46]
```diff
@@ -0,0 +1,63 @@
...
+@Injectable({
+  providedIn: 'root'
+})
+export class KittensService {
+
+  // List available Kittens. 
+  private AvailableKittens: Animal[] = [];
+  private jsonIsRead = false;
+
+  // List to manage user adopted  Kittens.
```
Comment (Benja-san) [inline src/app/services/kittens.service.ts]: remove dead code

File: src/app/services/kittens.service.ts [hunk 47]
```diff
@@ -0,0 +1,63 @@
...
+export class KittensService {
+
+  // List available Kittens. 
+  private AvailableKittens: Animal[] = [];
+  private jsonIsRead = false;
+
+  // List to manage user adopted  Kittens. 
+  private userAdoptedKittens: Animal[] = [];
+
+  // HttpClient Service injection to read json
```
Comment (Benja-san) [inline src/app/services/kittens.service.ts]: remove dead code

File: src/app/services/kittens.service.ts [hunk 48]
```diff
@@ -0,0 +1,63 @@
...
+  private http = inject(HttpClient)
+
+  // Json Read
```
Comment (Benja-san) [inline src/app/services/kittens.service.ts]: remove dead code

File: src/app/services/kittens.service.ts [hunk 49]
```diff
@@ -0,0 +1,63 @@
...
+  getJsonKittensAndPushToAvailableKittens(): void {
+    // if not read
+    if (!this.jsonIsRead) {
+      // Json to read :
```
Comment (Benja-san) [inline src/app/services/kittens.service.ts]: remove dead code

File: src/app/services/kittens.service.ts [hunk 50]
```diff
@@ -0,0 +1,63 @@
...
+import { Animal } from '../model/animal/animal';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class KittensService {
+
+  // List available Kittens. 
+  private AvailableKittens: Animal[] = [];
+  private jsonIsRead = false;
```
Comment (Benja-san) [inline src/app/services/kittens.service.ts]: start property by is if boolean use

**Review by Benja-san — CHANGES_REQUESTED** [review]

**Comment by carlsavila** [comment]
CR applied!

**Comment by carlsavila** [comment]
CR applied, dev/kittenservice conflicts resolved

**Review by Benja-san — APPROVED** [review]

**Pull request merged**